mutiny-core/src/lib.rs

```diff
@@ -876,6 +893,182 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+            let mut federation_storage_guard = self.federation_storage.lock().await;
+
+            if let Some(federation) = federation_storage_guard.federations.get_mut(uuid) {
+                federation.archived = true;
+
+                // Save the updated storage
+                self.storage
+                    .insert_federations(federation_storage_guard.clone())?;
+            } else {
+                return Err(MutinyError::NotFound);
+            }
+
+            // Remove the federation from the federations hashmap
+            federations_guard.remove(&federation_id);
+        } else {
+            return Err(MutinyError::NotFound);
+        }
+
+        Ok(())
+    }
+}
+
+async fn create_federation<S: MutinyStorage>(
+    storage: &S,
+    c: &MutinyWalletConfig,
+    logger: &Arc<MutinyLogger>,
+    stop: Arc<AtomicBool>,
+) -> Result<
+    (
+        FederationStorage,
+        Arc<Mutex<HashMap<FederationId, Arc<FederationClient<S>>>>>,
+    ),
+    MutinyError,
+> {
+    let federation_storage = storage.get_federations()?;
+    let unarchived_federations = federation_storage
+        .clone()
+        .federations
+        .into_iter()
+        .filter(|(_, n)| !n.is_archived());
+    let mut federation_map = HashMap::new();
+    for federation_item in unarchived_federations {
+        let federation = FederationClient::new(
+            federation_item.0,
+            &federation_item.1,
+            federation_item.1.federation_code.clone(),
+            c.xprivkey,
+            storage.clone(),
+            c.network,
+            logger.clone(),
+            stop.clone(),
+        )
+        .await?;
+
+        let id = federation.fedimint_client.federation_id();
+
+        federation_map.insert(id, Arc::new(federation));
+    }
+    let federations = Arc::new(Mutex::new(federation_map));
+    Ok((federation_storage, federations))
+}
+
+// This will create a new federation and returns the Federation ID of the client created.
+#[allow(clippy::too_many_arguments)]
+pub(crate) async fn create_new_federation<S: MutinyStorage>(
+    xprivkey: ExtendedPrivKey,
+    storage: S,
+    network: Network,
+    logger: Arc<MutinyLogger>,
+    federation_storage: Arc<Mutex<FederationStorage>>,
+    federations: Arc<Mutex<HashMap<FederationId, Arc<FederationClient<S>>>>>,
+    federation_code: InviteCode,
+    stop: Arc<AtomicBool>,
+) -> Result<FederationIdentity, MutinyError> {
+    // Begin with a mutex lock so that nothing else can
+    // save or alter the federation list while it is about to
+    // be saved.
+    let mut federation_mutex = federation_storage.lock().await;
+
+    // Get the current federations and their bip32 indices
+    // so that we can create another federation with the next.
+    // Always get it from our storage, the federation_mutex is
+    // mostly for read only and locking.
+    let mut existing_federations = storage.get_federations()?;
```

Comment:
should we check that the new federation isn't in our current list. Could see where you archive one and then add it back, we'd have it twice in the list, could just unarchive in that case, if it is active then can throw an error.